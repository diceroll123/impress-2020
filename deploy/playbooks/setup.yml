---
- name: Set up the environment for the impress-2020 app
  hosts: webserver
  vars:
    email_address: "emi@matchu.dev" # TODO: Extract this to personal config?
  tasks:
    - name: Create web user group
      become: yes
      group:
        name: web

    - name: Add current user to web group
      become: yes
      user:
        name: "{{ ansible_user_id }}"
        group: web
        append: yes

    - name: Create the app folder
      become: yes
      file:
        path: /srv/impress-2020
        state: directory
        # Root and the `web` group may read/write this folder. Everyone else
        # may only read it.
        group: web
        mode: "u=rwx,g=rwx,o=rx"

    - name: Add Nodesource apt key
      become: yes
      apt_key:
        id: 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280
        url: https://deb.nodesource.com/gpgkey/nodesource.gpg.key

    - name: Add Node v16 apt repository
      become: yes
      apt_repository:
        repo: deb https://deb.nodesource.com/node_16.x focal main

    - name: Install Node v16
      become: yes
      apt:
        update_cache: yes
        name: nodejs
        state: present

    - name: Install Yarn
      become: yes
      npm:
        name: yarn
        global: yes

    - name: Install pm2
      become: yes
      npm:
        name: pm2
        global: yes

    - name: Create pm2 startup script
      # The current user is going to become the pm2 owner of the app server
      # process. They'll be able to manage it without `sudo`, including during
      # normal deploys, and run `pm2 monit` from their shell to see status.
      become: yes
      command: "pm2 startup systemd {{ ansible_user_id }} --hp /home/{{ ansible_user_id }}"

    - name: Create pm2 ecosystem file
      copy:
        content: |
          module.exports = {
            apps: [
              {
                name: "impress-2020",
                cwd: "/srv/impress-2020/current",
                // Instead of `yarn start`, we specify the `next` binary
                // directly, because it helps pm2 monitor our app correctly.
                // https://github.com/vercel/next.js/discussions/10675#discussioncomment-34615
                script: "./node_modules/.bin/next",
                args: "start --port=3000",
                instances: "max",
                exec_mode: "cluster",
              }
            ]
          }
        dest: "~/ecosystem.config.js"
        # Create a temporary backup file, so we can use it to delete the old
        # version of the services. (This is important if e.g. a service is
        # removed or renamed, in which case deleting from the *new* config file
        # wouldn't include it.)
        backup: yes
      register: pm2_ecosystem_file

    - name: Delete old pm2 services if config file changed
      command: "pm2 delete {{ pm2_ecosystem_file.backup_file | quote }}"
      when: pm2_ecosystem_file is changed and pm2_ecosystem_file.backup_file is defined

    - name: Delete old pm2 config file if it changed
      file:
        path: "{{ pm2_ecosystem_file.backup_file }}"
        state: absent
      when: pm2_ecosystem_file is changed and pm2_ecosystem_file.backup_file is defined

    - name: Start pm2 services
      command: "pm2 start ~/ecosystem.config.js"

    - name: Save pm2 startup script
      command: pm2 save

    - name: Install core snap
      become: yes
      community.general.snap:
        name: core

    - name: Install certbot as a snap
      become: yes
      community.general.snap:
        name: certbot
        classic: yes

    - name: Set up certbot
      become: yes
      command: "certbot certonly --nginx -n --agree-tos --email {{ email_address }} --domains impress-2020-box.openneo.net"

    - name: Install nginx
      become: yes
      apt:
        update_cache: yes
        name: nginx

    - name: Add impress-2020 config file to nginx
      become: yes
      copy:
        content: |
          server {
            server_name impress-2020-box.openneo.net;
            listen 80;
            if ($host = impress-2020-box.openneo.net) {
              return 301 https://$host$request_uri;
            }
          }

          server {
            server_name impress-2020-box.openneo.net;
            listen 443 ssl;
            ssl_certificate /etc/letsencrypt/live/impress-2020-box.openneo.net/fullchain.pem;
            ssl_certificate_key /etc/letsencrypt/live/impress-2020-box.openneo.net/privkey.pem;
            include /etc/letsencrypt/options-ssl-nginx.conf;
            ssl_dhparam /etc/letsencrypt/ssl-dhparams.pem;
            ssl_session_cache shared:SSL:10m; # https://superuser.com/q/1484466/14127

            # TODO: Serve static files directly, instead of through the proxy
            location / {
              proxy_pass http://127.0.0.1:3000;
            }
          }
        dest: /etc/nginx/sites-enabled/impress-2020
      notify:
        - Restart nginx

    - name: Install dependencies for the npm module node-canvas
      become: yes
      apt:
        update_cache: yes
        name:
          - build-essential
          - libcairo2-dev
          - libpango1.0-dev
          - libjpeg-dev
          - libgif-dev
          - librsvg2-dev

  handlers:
    - name: Restart nginx
      become: yes
      systemd:
        name: nginx
        state: restarted
