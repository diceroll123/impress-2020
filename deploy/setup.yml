---
- name: Install the impress-2020 app
  hosts: webserver
  tasks:
    - name: Create web user group
      become: yes
      group:
        name: web
    - name: Add current user to web group
      become: yes
      user:
        name: "{{ ansible_user_id }}"
        group: web
        append: yes
    - name: Create the app folder
      become: yes
      file:
        path: /srv/impress-2020
        state: directory
        # Root and the `web` group may read/write this folder. Everyone else
        # may only read it.
        group: web
        mode: "u=rwx,g=rwx,o=rx"
    - name: Add Nodesource apt key
      become: yes
      apt_key:
        id: 9FD3B784BC1C6FC31A8A0A1C1655A0AB68576280
        url: https://deb.nodesource.com/gpgkey/nodesource.gpg.key
    - name: Add Node v16 apt repository
      become: yes
      apt_repository:
        repo: deb https://deb.nodesource.com/node_16.x focal main
    - name: Install Node v16
      become: yes
      apt:
        update_cache: yes
        name: nodejs
        state: present
