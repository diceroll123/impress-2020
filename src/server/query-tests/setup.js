const fs = require("fs");
const path = require("path");

const { ApolloServer } = require("apollo-server");
const { createTestClient } = require("apollo-server-testing");
const { AuthenticationClient } = require("auth0");

const connectToDb = require("../db");
const actualConnectToDb = jest.requireActual("../db");
const { config } = require("../index");

let accessTokenForQueries = null;

const { query } = createTestClient(
  new ApolloServer({
    ...config,
    context: () =>
      config.context({
        req: {
          headers: {
            authorization: accessTokenForQueries
              ? `Bearer ${accessTokenForQueries}`
              : undefined,
          },
        },
      }),
  })
);

// Spy on db.execute, so we can snapshot the queries we run. This can help us
// keep an eye on perf - watch for tests with way too many queries!
jest.mock("../db");
let dbExecuteFn;
let db;
let dbEnvironment = "production";
let dbSetupDone = false;
const dbSetupScripts = [
  fs
    .readFileSync(
      path.join(__dirname, "../../../scripts/setup-mysql-dev-constants.sql")
    )
    .toString(),
  fs
    .readFileSync(
      path.join(__dirname, "../../../scripts/setup-mysql-dev-schema.sql")
    )
    .toString(),
];
beforeAll(() => {
  connectToDb.mockImplementation(async () => {
    let options;

    if (dbEnvironment === "test") {
      options = {
        host: "localhost",
        user: "impress_2020_test",
        password: "impress_2020_test",
        database: "impress_2020_test",
      };
    }

    db = await actualConnectToDb(options);

    if (dbEnvironment === "test" && !dbSetupDone) {
      for (const script of dbSetupScripts) {
        await db.query(script);
      }
    }
    dbSetupDone = true;

    dbExecuteFn = jest.spyOn(db, "execute");
    return db;
  });

  // Mock out a current "now" date, for consistent snapshots
  const NOW = new Date("2020-01-01T00:00:00.000Z");
  jest.spyOn(global, "Date").mockImplementation(() => NOW);
});
beforeEach(() => {
  accessTokenForQueries = null;
  if (dbExecuteFn) {
    dbExecuteFn.mockClear();
  }
  dbEnvironment = "production";
  dbSetupDone = false;
  db = null;
});
afterAll(() => {
  if (db) {
    db.end();
  }
  Date.mockRestore();
});
const getDbCalls = () => (dbExecuteFn ? dbExecuteFn.mock.calls : []);
const clearDbCalls = () => dbExecuteFn?.mockClear();

function useTestDb() {
  if (db) {
    throw new Error(`can't call useTestDb() if db mock already exists`);
  }
  dbEnvironment = "test";
}

async function logInAsTestUser() {
  const auth0 = new AuthenticationClient({
    domain: "openneo.us.auth0.com",
    clientId: process.env.AUTH0_TEST_CLIENT_ID,
    clientSecret: process.env.AUTH0_TEST_CLIENT_SECRET,
  });

  const res = await auth0.passwordGrant({
    username: "dti-test",
    password: process.env.DTI_TEST_USER_PASSWORD,
    audience: "https://impress-2020.openneo.net/api",
  });

  accessTokenForQueries = res.access_token;
}

// Add a new `expect(res).toHaveNoErrors()` to call after GraphQL calls!
expect.extend({
  toHaveNoErrors(res) {
    if (res.errors) {
      return {
        message: () =>
          `expected no GraphQL errors, but got:\n    ${res.errors}`,
        pass: false,
      };
    } else {
      return {
        message: () => `expected GraphQL errors, but there were none`,
        pass: true,
      };
    }
  },
});

// Use the new modeling code, even though it's disabled in most environments,
// in order to test it.
process.env["USE_NEW_MODELING"] = "1";

module.exports = {
  query,
  getDbCalls,
  clearDbCalls,
  connectToDb,
  useTestDb,
  logInAsTestUser,
};
