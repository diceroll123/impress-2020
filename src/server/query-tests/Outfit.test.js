import gql from "graphql-tag";
import { query, getDbCalls } from "./setup.js";

describe("Outfit", () => {
  it("loads an outfit by ID", async () => {
    const res = await query({
      query: gql`
        query {
          outfit(id: "31856") {
            id
            name

            petAppearance {
              id

              color {
                id
                name
              }

              species {
                id
                name
              }

              pose
            }

            wornItems {
              id
              name
            }

            closetedItems {
              id
              name
            }
          }
        }
      `,
    });

    expect(res).toHaveNoErrors();
    expect(res.data).toMatchSnapshot();
    expect(getDbCalls()).toMatchInlineSnapshot(`
      Array [
        Array [
          "SELECT * FROM outfits WHERE id IN (?)",
          Array [
            "31856",
          ],
        ],
        Array [
          "SELECT * FROM item_outfit_relationships WHERE outfit_id IN (?)",
          Array [
            "31856",
          ],
        ],
        Array [
          "SELECT * FROM pet_states WHERE id IN (?)",
          Array [
            "3951",
          ],
        ],
        Array [
          "SELECT * FROM item_translations WHERE item_id IN (?,?,?,?,?,?,?,?,?,?,?) AND locale = \\"en\\"",
          Array [
            "38916",
            "51054",
            "38914",
            "36125",
            "36467",
            "47075",
            "47056",
            "39662",
            "56706",
            "38915",
            "56398",
          ],
        ],
        Array [
          "SELECT * FROM pet_types WHERE id IN (?)",
          Array [
            "33",
          ],
        ],
        Array [
          "SELECT * FROM color_translations
             WHERE color_id IN (?) AND locale = \\"en\\"",
          Array [
            "34",
          ],
        ],
        Array [
          "SELECT * FROM species_translations
             WHERE species_id IN (?) AND locale = \\"en\\"",
          Array [
            "54",
          ],
        ],
      ]
    `);
  });
});
