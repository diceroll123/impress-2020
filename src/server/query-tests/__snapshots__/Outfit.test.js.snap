// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Outfit loads an outfit by ID 1`] = `
Object {
  "outfit": Object {
    "closetedItems": Array [
      Object {
        "id": "36125",
        "name": "Blue Newsboy Hat",
      },
      Object {
        "id": "36467",
        "name": "Daring Adventurer Hat",
      },
      Object {
        "id": "47075",
        "name": "Jordies Adventure Hat",
      },
      Object {
        "id": "47056",
        "name": "Moltara Inventor Hat and Goggles",
      },
      Object {
        "id": "39662",
        "name": "Simple Sun Hat",
      },
      Object {
        "id": "56706",
        "name": "Super Sleuth Hat and Wig",
      },
      Object {
        "id": "38915",
        "name": "Zafara Tourist Shirt",
      },
      Object {
        "id": "56398",
        "name": "Altador Cup Kreludor Frame",
      },
    ],
    "id": "31856",
    "name": "Zafara Tourist",
    "petAppearance": Object {
      "color": Object {
        "id": "34",
        "name": "Green",
      },
      "id": "3951",
      "pose": "UNKNOWN",
      "species": Object {
        "id": "54",
        "name": "Zafara",
      },
    },
    "wornItems": Array [
      Object {
        "id": "38916",
        "name": "Zafara Tourist Camera",
      },
      Object {
        "id": "51054",
        "name": "Summer Fun Beach Background",
      },
      Object {
        "id": "38914",
        "name": "Zafara Tourist Hat",
      },
    ],
  },
}
`;
